pyrsistent/__init__.pyi

```diff
@@ -18,77 +19,97 @@ from typing import Type
 from typing import TypeVar
 from typing import overload
 
+from pyrsistent.typing import CheckedKeyTypeError as CheckedKeyTypeError
+from pyrsistent.typing import CheckedPMap as CheckedPMap
+from pyrsistent.typing import CheckedPSet as CheckedPSet
+from pyrsistent.typing import CheckedPVector as CheckedPVector
+from pyrsistent.typing import CheckedType as CheckedType
+from pyrsistent.typing import CheckedValueTypeError as CheckedValueTypeError
+from pyrsistent.typing import InvariantException as InvariantException
+from pyrsistent.typing import PClass as PClass
 from pyrsistent.typing import PBag as PBag
 from pyrsistent.typing import PDeque as PDeque
 from pyrsistent.typing import PList as PList
 from pyrsistent.typing import PMap as PMap
 from pyrsistent.typing import PMapEvolver as PMapEvolver
 from pyrsistent.typing import PSet as PSet
 from pyrsistent.typing import PSetEvolver as PSetEvolver
+from pyrsistent.typing import PTypeError as PTypeError
 from pyrsistent.typing import PVector as PVector
 from pyrsistent.typing import PVectorEvolver as PVectorEvolver
 
+T = TypeVar('T')
 KT = TypeVar('KT')
 VT = TypeVar('VT')
 
-def pmap(initial: Mapping[KT, VT] = {}, pre_size: int = 0) -> PMap[KT, VT]: ...
-def m(**kwargs: Mapping[KT, VT]) -> PMap[KT, VT]: ...
-
-T = TypeVar('T')
+def pmap(initial: Union[Mapping[KT, VT], Iterable[Tuple[KT, VT]]] = {}, pre_size: int = 0) -> PMap[KT, VT]: ...
+def m(**kwargs: VT) -> PMap[str, VT]: ...
 
-def pvector(iterable: Iterable[T]) -> PVector[T]: ...
-def v(*iterable: Iterable[T]) -> PVector[T]: ...
+def pvector(iterable: Iterable[T] = ()) -> PVector[T]: ...
+def v(*iterable: T) -> PVector[T]: ...
 
 def pset(iterable: Iterable[T] = (), pre_size: int = 8) -> PSet[T]: ...
-def s(*iterable: Iterable[T]) -> PSet[T]: ...
+def s(*iterable: T) -> PSet[T]: ...
+
+# see class_test.py for use cases
+Invariant = Tuple[bool, Optional[Union[str, Callable[[], str]]]]
 
 # The actual return value (_PField) is irrelevant after a PRecord has been instantiated,
 # see https://github.com/tobgu/pyrsistent/blob/master/pyrsistent/_precord.py#L10
 def field(
-    type: Union[Type[T], Sequence[Type[T]]] = (),
-    invariant: Callable[[Any], Tuple[bool, Optional[str]]] = lambda _: (True, None),
+    type: Union[str, Type[T], Sequence[Type[T]]] = (),
+    invariant: Callable[[Any], Union[Invariant, Iterable[Invariant]]] = lambda _: (True, None),
     initial: Any = object(),
     mandatory: bool = False,
     factory: Callable[[Any], T] = lambda x: x,
     serializer: Callable[[Any, T], Any] = lambda _, value: value,
 ) -> T: ...
 
+# This is not type safe, probably requires this:
```

Comment:
Yeah there probably isn't a way to utilize the stringy types, but what i had in mind when writing the comment was something similar to this: https://github.com/python/typeshed/blob/598d0d6/stdlib/2and3/builtins.pyi#L1184
I figured it would give precise types for declarations like `item_type=(int, str, bytes)` as long as we had the overload for tuples with length of three.
I did some testing and looks like this is probably not a good idea, mainly because we cannot represent tuple of at least n items, additionally it looks pretty nasty because the definitions are copy-pasted many times.

I'll push an alternative approach for this in a moment, see what you think about that.